patroni/postgresql.py

```diff
@@ -733,6 +736,12 @@ def get_pid(self):
             logger.warning("Garbage pid in postmaster.pid: {0!r}".format(pid))
             return 0
 
+    def get_pid_with_lost_data_dir(self):
+        process = list(filter(lambda p: p.name() == "postgres" and self._data_dir in p.cmdline(), psutil.process_iter()))
```

Comment:
Unfortunately this check is not really safe: `self._data_dir in p.cmdline()`

If you have more than one postgres cluster running on the same host with some strange configuration, it might result that more than one postgres would be killed.

The best we could do - keep the content of `postmaster.pid` in memory (especially lines 1 and 3, containing postmaster pid and start_time), and if data directory has gone away, use this cached content to terminate postmaster.

Reply:
I was just thinking along similar lines about whether we should create a `psutil.Process` object and cache that. This will do the hard work of checking if we still have the same process running. I think it might even simplify the code a good bit.

Reply:
Looking at psutil source, Process class just determines identity via `(pid, create_time)`, so rolling our own would also work. But as we need the psutil dependency anyway for `_wait_for_user_backends_to_close()` we might as well use it here too. Seems like a pretty well tested package.

Reply:
Changed logic so it get postmaster from PID file and then during stop compares cached start_time from postmaster and create_time from psutls. Actually they are slightly different so I am checking difference between them.